projects/observability/src/shared/components/cartesian/d3/legend/cartesian-legend.ts

```diff
@@ -202,16 +260,57 @@ export class CartesianLegend {
     this.activeSeriesSubject.next();
   }
 
-  private updateActiveSeries(seriesEntry: Series<{}>): void {
-    if (this.isDefault) {
-      this.activeSeries = [];
-      this.activeSeries.push(seriesEntry);
-      this.isDefault = false;
+  private getGroupedSeries(): Series<{}>[][] {
+    if (!this.isGrouped) {
+      return [];
+    }
+
+    const nonHiddenSeries = this.series.filter(series => !series.hide);
+    const groupedSeries: Series<{}>[][] = [];
+    const currentGroup: Series<{}>[] = [];
+
+    nonHiddenSeries.forEach(seriesEntry => {
+      if (
+        currentGroup.length === 0 ||
+        currentGroup[currentGroup.length - 1].assignedGroup!.id === seriesEntry.assignedGroup!.id
+      ) {
+        currentGroup.push(seriesEntry);
+      } else {
+        groupedSeries.push([...currentGroup]);
+        currentGroup.length = 0;
+        currentGroup.push(seriesEntry);
+      }
+    });
+    groupedSeries.push([...currentGroup]);
+
+    return groupedSeries;
+  }
+
+  private updateActiveSeries(series: Series<{}> | Series<{}>[]): void {
```

Comment:
If we can possibly keep the data consistent (e.g. always use arrays, even if ungrouped), that would help keep the code cleaner and avoid things like instanceof.

Reply:
I thought of doing the same earlier, but I ran into one problem.
Since we have whole group selection thing as well, that creates an issue. Suppose for non-grouped case what I would do is, I will pass [seriesA] as an argument (If I go with array approach), now what I would do is, foe every entry present in the array, I would check whether it is active or not , which works well with non-grouped case.

Now since we have whole group selection this time, and if I pass [A, B, C, D, E] with already D and E active then the above mentioned approach would make A, B, C => active and D, E inactive. SO I needed a way to distinguish between whole group selection or individual one. I can use another argument to do that, which can remove that instance of atleast, but it might extra handling of the code inside.

Reply:
OK, so given the additional requirement of toggling a whole series at once, how do we expect it to behave? In the scenario you described, would we want A,B,C to be set to active and D,E not to change? Or D,E to set to inactive and A,B,C not to change? That resolution logic of a grouped series could be it own thing, but my suggestion would be that this takes in our top level unit - (which, in my other recommendation would be a `SeriesGroup`, even if that `SeriesGroup` contains one).

In other words, we do want to be able to make the determination "what got clicked", but that's not at odds with keeping the data shape consistent. The right abstraction (which doesn't sound like it should be just an array of series), would allow that too and address the consistency problem.